models/glif_lif_asc.h

```diff
@@ -0,0 +1,285 @@
+coefficients (fj in Equation (7) in [1]).
+  V_dynamics_method string - Voltage dynamics (Equation (1) in [1]) solution
+methods:
+                             'linear_forward_euler' - Linear Euler forward (RK1)
+to find next V_m value, or
+                             'linear_exact' - Linear exact to find next V_m
+value.
+
+References:
+  [1] Teeter C, Iyer R, Menon V, Gouwens N, Feng D, Berg J, Szafer A,
+      Cain N, Zeng H, Hawrylycz M, Koch C, & Mihalas S (2018)
+      Generalized leaky integrate-and-fire models classify multiple neuron
+types.
+      Nature Communications 9:709.
+
+Author: Binghuang Cai and Kael Dai @ Allen Institute for Brain Science
+*/
+
+namespace nest
+{
+
+class glif_lif_asc : public nest::Archiving_Node
+{
+public:
+  glif_lif_asc();
+
+  glif_lif_asc( const glif_lif_asc& );
+
+  using nest::Node::handle;
+  using nest::Node::handles_test_event;
+
+  nest::port send_test_event( nest::Node&, nest::port, nest::synindex, bool );
+
+  void handle( nest::SpikeEvent& );
+  void handle( nest::CurrentEvent& );
+  void handle( nest::DataLoggingRequest& );
+
+  nest::port handles_test_event( nest::SpikeEvent&, nest::port );
+  nest::port handles_test_event( nest::CurrentEvent&, nest::port );
+  nest::port handles_test_event( nest::DataLoggingRequest&, nest::port );
+
+  bool is_off_grid() const // uses off_grid events
+  {
+    return true;
+  }
+
+  void get_status( DictionaryDatum& ) const;
+  void set_status( const DictionaryDatum& );
+
+private:
+  //! Reset parameters and state of neuron.
+
+  //! Reset state of neuron.
+  void init_state_( const Node& proto );
+
+  //! Reset internal buffers of neuron.
+  void init_buffers_();
+
+  //! Initialize auxiliary quantities, leave parameters and state untouched.
+  void calibrate();
+
+  //! Take neuron through given time interval
+  void update( nest::Time const&, const long, const long );
+
+  // The next two classes need to be friends to access the State_ class/member
+  friend class nest::RecordablesMap< glif_lif_asc >;
+  friend class nest::UniversalDataLogger< glif_lif_asc >;
+
+
+  struct Parameters_
+  {
+    double V_th_;    // A constant spiking threshold in mV
+    double G_;       // membrane conductance in nS
+    double E_L_;     // resting potential in mV
+    double C_m_;     // capacitance in pF
+    double t_ref_;   // refractory time in ms
+    double V_reset_; // Membrane voltage following spike in mV
+
+    std::vector< double > asc_init_; // initial values of ASCurrents_ in pA
+    std::vector< double > k_;        // predefined time scale in 1/ms
+    std::vector< double > asc_amps_; // in pA
+    std::vector< double > r_;        // coefficient
+    std::string V_dynamics_method_;  // voltage dynamic methods
+
+    Parameters_();
+
+    void get( DictionaryDatum& ) const;
+    void set( const DictionaryDatum& );
+  };
+
+
+  struct State_
+  {
+    double V_m_;                       // membrane potential in mV
```

Comment:
In almost all other linear neuron models, we work internally with `U = V_m - E_rev`. This makes the numerics slightly simpler and avoids any problems when changing `E_rev`. The `State::get()/set()` methods convert between `U` and `V_m`.

Reply:
Models have been updated to use relative membrane potential internally. We will update this branch shortly.